### ISSUE TITLE
Resuming Training

### ISSUE THREAD

**Issue by xamm** [huggingface/accelerate#95]
My `model` and `optimizer` are prepared using `.prepare()`,
my `scheduler` is left without "preparation" as was done in the [nlp_example.py](https://github.com/huggingface/accelerate/blob/e93cb7a3bd8420c07f28de034e49922e4e604d19/examples/nlp_example.py)

If I want to resume training, do I need to use the `.prepare()` function again after loading the previously saved model?

Or would I need to use the `.prepare()` function only if I loaded the trained model before calling `.prepare()`?

**Comment by sgugger** [comment]
You only need to use `prepare` once. If you saved the unwrapped model given by the `accelerator.unwrap_model(model)` method, you should first load your model then send it to `prepare`. If you saved the wrapped model, you should first send it to `prepare` then load your checkpoint.

**Comment by xamm** [comment]
Thanks for your answer.

So there is a different order for calling prepare and loading the checkpoint, depending on whether the model has been unpacked or not?

**Comment by sgugger** [comment]
Yes, since the `prepare` method will often put it in a wrapper (like `DistributedDataParallel`).

**Comment by TPFRL** [comment]
@sgugger Could you explain the suggested way to save AcceleratedOptimizer??

In Huggingface Trainer, I can see that they use torch_xla.core.xla_model.save if using TPU.
https://github.com/huggingface/transformers/blob/2e5dbdf2db4599a6694d0974575a70f9bc3c978e/src/transformers/trainer.py#L1486-L1491


Would it be safe to simply use like the following??

if accelerator.is_local_main_process:
  torch.save({"opt": optimizer.state_dict()}, 'save.pkl')

Then,

optimizer.load_state_dict('save.pkl')

**Comment by sgugger** [comment]
You should just use `accelerator.save` instead of `torch.save` and remove the test: `accelerator.save` will only save on the main process and it will perform that special code necessary on TPUs.

**Comment by gowtham1997** [comment]
Hello @sgugger.
For fp16 training, im additionally saving the scalar like this:
```python
if args.output_dir is not None:
        accelerator.wait_for_everyone()
        unwrapped_model = accelerator.unwrap_model(model)
        unwrapped_model.save_pretrained(
            os.path.join(args.output_dir, "last_model"), save_function=accelerator.save
        )
        accelerator.save(
            {
                "epoch": completed_epochs,
                "steps": completed_steps,
                "optimizer": optimizer.state_dict(),
                "scheduler": lr_scheduler.state_dict(),
                "scaler": accelerator.scaler.state_dict(),
            },
            args.optim_scheduler_states_path,
        )
```

This is how i load everything back for resuming training:

```python
    param = DistributedDataParallelKwargs(
            find_unused_parameters=True, check_reduction=False
        )

    # Initialize the accelerator. We will let the accelerator handle device placement for us in this example.
    accelerator = Accelerator(fp16=args.fp16, kwargs_handlers=[param])
    
    # setting up dataset, tokenization, dataloader, model, optimizer, etc
    ...
    ...
    
    
    if args.resume_from_checkpoint_dir is not None:
        logger.info("Loading model weights from checkpoint ..")
        # we load the model before preparing
        # see this for details: https://github.com/huggingface/accelerate/issues/95
        model.from_pretrained(args.resume_from_checkpoint_dir)

        optim_scheduler_states = torch.load(args.optim_scheduler_states_path)

        logger.info("Loading optimizer states from checkpoint dir ..")
        accelerator.scaler.load_state_dict(optim_scheduler_states["scaler"])
        optimizer.load_state_dict(optim_scheduler_states["optimizer"])
        completed_epochs = optim_scheduler_states["epoch"]
        completed_steps = optim_scheduler_states["steps"]
        lr_scheduler.load_state_dict(optim_scheduler_states["scheduler"])

    # Prepare everything with our `accelerator`.
    model, optimizer, train_dataloader, eval_dataloader = accelerator.prepare(
        model, optimizer, train_dataloader, eval_dataloader
    )
    
    if (
        accelerator.distributed_type == DistributedType.MULTI_GPU
        or accelerator.distributed_type == DistributedType.TPU
    ):
        # taken from https://github.com/pytorch/fairseq/blob/a8c8f0be177649f8178fdcbae519c17894efd4d7/fairseq/distributed/module_proxy_wrapper.py#L9
        # forward missing getattr and state_dict/load_state_dict to orig model
        model = ModuleProxyWrapper(model)
    
```

But the loaded model starts out with a high loss and seems like it's starting fresh. Do you see any bugs in the snippet?

**Comment by sgugger** [comment]
It does look like you have thought of everything (you even have the loss scaler properly saved) so it looks correct to me. How is the reported loss computed? Maybe there is an average problem when you resume training?

**Comment by gowtham1997** [comment]
Thanks for the prompt reply.

Most of my code is just a modified version of hf [train_mlm_no_trainer] example (https://github.com/huggingface/transformers/blob/master/examples/pytorch/language-modeling/run_mlm_no_trainer.py) 

Here is the validation snippet, both the val loss and the mlm_accuracy seem to indicate the model is starting fresh

```python
def validate_model(
    model,
    eval_dataloader,
    accelerator,
    len_eval_dataset,
    per_device_eval_batch_size,
    pad_to_max_length,
):
    metric = load_metric("custom_metrics/mlm_accuracy.py")

    def get_labels(predictions, references):
        # Transform predictions and references tensos to numpy arrays
        if accelerator.device.type == "cpu":
            y_pred = predictions.detach().clone().numpy()
            y_true = references.detach().clone().numpy()
        else:
            y_pred = predictions.detach().cpu().clone().numpy()
            y_true = references.detach().cpu().clone().numpy()

        # Remove ignored index (special tokens)
        true_predictions = [
            [str(p) for (p, l) in zip(pred, gold_label) if l != -100]
            for pred, gold_label in zip(y_pred, y_true)
        ]
        true_labels = [
            [str(l) for (p, l) in zip(pred, gold_label) if l != -100]
            for pred, gold_label in zip(y_pred, y_true)
        ]

        return true_predictions, true_labels

    losses = []
    model.eval()
    for step, batch in enumerate(eval_dataloader):
        # We could avoid this line since we set the accelerator with `device_placement=True`.
        batch.to(accelerator.device)
        with torch.no_grad():
            outputs = model(**batch)

        loss = outputs.loss
        losses.append(accelerator.gather(loss.repeat(per_device_eval_batch_size)))

        predictions = outputs.logits.argmax(dim=-1)
        labels = batch["labels"]
        if (
            not pad_to_max_length
        ):  # necessary to pad predictions and labels for being gathered
            predictions = accelerator.pad_across_processes(
                predictions, dim=1, pad_index=-100
            )
            labels = accelerator.pad_across_processes(labels, dim=1, pad_index=-100)
        predictions_gathered = accelerator.gather(predictions)
        labels_gathered = accelerator.gather(labels)

        preds, refs = get_labels(predictions_gathered, labels_gathered)
        metric.add_batch(
            predictions=preds,
            references=refs,
        )  # predictions and preferences are expected to be a nested list of labels, not label_ids

    losses = torch.cat(losses)
    losses = losses[:len_eval_dataset]
    eval_metric = metric.compute()

    try:
        val_loss = torch.mean(losses)
        perplexity = math.exp(torch.mean(losses))
    except OverflowError:
        perplexity = float("inf")

    eval_metric["val_loss"] = val_loss
    eval_metric["perplexity"] = perplexity

    return eval_metric
```

**Comment by Yaicky** [comment]
Hi, @sgugger .
I find one accelerator.gather() issue related to saving method in distributed evaluation.
if I  `accelerator.save(wrapped_model)`, and for loading, `prepare`, `load(wrapped_model_ckpt)`, just like you said before on this issue page, when I appended `accelerator.gather()`(same codes in cv_example.py) to my eval function, I got stuck at the **_second_** iteration, no errors, no exits, just stuck(I use tqdm, progressbar didn't go.)

So I followed your official document to save unwrapped_model first, thing goes well.
Maybe it's not an issue cause saving unwrapped_model is recommended, just FYI. :)

**Comment by xamm** [comment]
So my resuming pipeline is as follows:

Saving the model in unwrapped state:

        self.__accelerator.wait_for_everyone()
        unwrapped = self.__accelerator.unwrap_model(self.model)

        unwrapped.save_pretrained(out_path, save_function=self.__accelerator.save, state_dict=self.__accelerator.get_state_dict(unwrapped))

        self.__accelerator.save(
            {
                'model': unwrapped.state_dict(),
                'optimizer': self.optim.state_dict(),
                'scheduler': self.lr_scheduler.state_dict(),
                'scaler': self.__accelerator.scaler.state_dict(),

            },
            os.path.join(out_path, "model.pt"),
        )

Loading the model:

```
    ckpoint = torch.load(model_file)
    self.model.from_pretrained(input)
    self.optim.load_state_dict(ckpoint["optimizer"])
    self.lr_scheduler.load_state_dict(ckpoint["scheduler"])
```

Afterwards send model, optimizer and dataloaders through `prepare`.

As far as I understand this should work, but for me the same problem which @Yaicky mentions occurs. `.gather()` is stuck in evaluation and training mode.

**Comment by sgugger** [comment]
Are you sure the tensors you are gathering have the same sizes? I remember getting the processes stuck with no error message during evaluation if this was not the case.

**Comment by gowtham1997** [comment]
@sgugger
With the same code as before
```python
# this didnt work (not sure why)
 model.from_pretrained(args.resume_from_checkpoint_dir)
 
# replaced with this and the metrics are proper and i can resume training
model.load_state_dict(torch.load(os.path.join(args.resume_from_checkpoint_dir, "pytorch_model.bin")))
 
```

**Comment by xamm** [comment]
> Are you sure the tensors you are gathering have the same sizes? I remember getting the processes stuck with no error message during evaluation if this was not the case.

The tensors have the same size. It really seems that the gather() function never returns. When I remove the gather calls, the model runs fine.
But my loss is kept at the same level when the model was stopped. So it appears that no learning occurs after resumption. But the model is in train mode.

**Comment by BiEchi** [comment]
Dear @sgugger : if I `prepare` then `unwrap` a model before loading it with `self.accelerator.load_state()`, will it still get loaded?
